src/Analyzers/CSharp/Analyzers/SimplifyLinqExpression/CSharpSimplifyLinqExpressionDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,179 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Operations;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Roslyn.Utilities;
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+#if CODE_STYLE
+using OptionSet = Microsoft.CodeAnalysis.Diagnostics.AnalyzerConfigOptions;
+#endif
+
+namespace Microsoft.CodeAnalysis.CSharp.SimplifyLinqExpression
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal sealed class CSharpSimplifyLinqExpressionDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        private readonly ImmutableArray<string> _nonEnumerableReturningLinqMethodNames =
+            ImmutableArray.Create(
+                nameof(Enumerable.First),
+                nameof(Enumerable.Last),
+                nameof(Enumerable.Single),
+                nameof(Enumerable.Any),
+                nameof(Enumerable.Count),
+                nameof(Enumerable.SingleOrDefault),
+                nameof(Enumerable.FirstOrDefault),
+                nameof(Enumerable.LastOrDefault)
+            );
+
+        public CSharpSimplifyLinqExpressionDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.SimplifyLinqExpressionsDiagnosticId,
+                   EnforceOnBuildValues.SimplyLinq,
+                   option: null,
+                   title: new LocalizableResourceString(
+                       nameOfLocalizableResource: nameof(CSharpAnalyzersResources.Simplify_Linq_expression),
+                       resourceManager: CSharpAnalyzersResources.ResourceManager,
+                       resourceSource: typeof(CSharpAnalyzersResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            var whereMethods = ImmutableHashSet.CreateBuilder<IMethodSymbol>();
+            var linqMethods = ImmutableHashSet.CreateBuilder<IMethodSymbol>();
```

Comment:
immutable hashing collections are incredible slow and memory wasteful.  do we need this here?  if so, why.